File: workers/src/lib/capture.ts
import type { CaptureEntry, CourtOption, Params, RequestData, TimeSlot, VenueSnapshot } from "../types";
import { parseDate } from "./utils";

const SUBMIT_PATH = "/v2/reserve/submit?";
const SITE_LIST_PREFIX = "/v1/venues/venues_site_list";
const CALENDAR_PREFIX = "/v1/venues/calendar";

export const FIXED_COURTS: CourtOption[] = [
  { site_id: 3692729935134806, site_name: "1号场" },
  { site_id: 3692729935134807, site_name: "2号场" },
  { site_id: 3692729935134808, site_name: "3号场" },
  { site_id: 3692729935134809, site_name: "4号场" },
  { site_id: 3692729935134810, site_name: "5号场" },
  { site_id: 3692729935134811, site_name: "6号场" },
  { site_id: 3704542185717805, site_name: "7号场" },
  { site_id: 3713773110689834, site_name: "6楼V1号场" },
  { site_id: 3713764789190689, site_name: "6楼V2号场" },
];

export const FIXED_SOURCE_DATE = "2026/05/28";

export const FIXED_TIME_SLOTS: TimeSlot[] = [
  { start_time: "07:00", end_time: "08:00", start_timestamp: 1779922800, end_timestamp: 1779926400, price: "75", times: "1" },
  { start_time: "08:00", end_time: "09:00", start_timestamp: 1779926400, end_timestamp: 1779930000, price: "75", times: "1" },
  { start_time: "09:00", end_time: "10:00", start_timestamp: 1779930000, end_timestamp: 1779933600, price: "75", times: "1" },
  { start_time: "10:00", end_time: "11:00", start_timestamp: 1779933600, end_timestamp: 1779937200, price: "75", times: "1" },
  { start_time: "11:00", end_time: "12:00", start_timestamp: 1779937200, end_timestamp: 1779940800, price: "75", times: "1" },
  { start_time: "12:00", end_time: "13:00", start_timestamp: 1779940800, end_timestamp: 1779944400, price: "75", times: "1" },
  { start_time: "13:00", end_time: "14:00", start_timestamp: 1779944400, end_timestamp: 1779948000, price: "75", times: "1" },
  { start_time: "14:00", end_time: "15:00", start_timestamp: 1779948000, end_timestamp: 1779951600, price: "130", times: "1" },
  { start_time: "15:00", end_time: "16:00", start_timestamp: 1779951600, end_timestamp: 1779955200, price: "130", times: "1" },
  { start_time: "16:00", end_time: "17:00", start_timestamp: 1779955200, end_timestamp: 1779958800, price: "130", times: "1" },
  { start_time: "17:00", end_time: "18:00", start_timestamp: 1779958800, end_timestamp: 1779962400, price: "130", times: "1" },
  { start_time: "18:00", end_time: "19:00", start_timestamp: 1779962400, end_timestamp: 1779966000, price: "130", times: "1" },
  { start_time: "19:00", end_time: "20:00", start_timestamp: 1779966000, end_timestamp: 1779969600, price: "130", times: "1" },
  { start_time: "20:00", end_time: "21:00", start_timestamp: 1779969600, end_timestamp: 1779973200, price: "130", times: "1" },
  { start_time: "21:00", end_time: "22:00", start_timestamp: 1779973200, end_timestamp: 1779976800, price: "130", times: "1" },
];

export class CaptureStore {
  entries: CaptureEntry[];

  constructor(entries: CaptureEntry[]) {
    this.entries = entries;
  }

  static async fromKV(kv: KVNamespace, key = "request_template:default"): Promise<CaptureStore> {
    const raw = await kv.get(key);
    if (!raw) throw new Error("No request template in KV. Upload request.txt first.");
    return new CaptureStore(JSON.parse(raw));
  }

  submitEntry(): CaptureEntry {
    const entry = this.entries.find((e) => e.path === SUBMIT_PATH);
    if (!entry) throw new Error(`No request found for path: ${SUBMIT_PATH}`);
    return entry;
  }

  submitBody(): Record<string, any> {
    return decodeJsonBody(this.submitEntry().req);
  }

  submitHeaders(): Record<string, string> {
    const headers = { ...this.submitEntry().req.headers };
    delete headers["host"];
    delete headers["content-length"];
    delete headers["accept-encoding"];
    headers["content-type"] = "application/json";
    return headers;
  }

  latestSiteListEntry(): CaptureEntry | null {
    const entries = this.entries.filter(
      (e) => e.path?.startsWith(SITE_LIST_PREFIX) && e.hostname === "stmember.styd.cn"
    );
    if (!entries.length) return null;
    return entries.reduce((a, b) => ((a.order || 0) > (b.order || 0) ? a : b));
  }

  siteListEntry(): CaptureEntry {
    const entry = this.latestSiteListEntry();
    if (!entry) throw new Error("No venues_site_list request found");
    return entry;
  }

  siteListHeaders(): Record<string, string> {
    const headers = { ...this.siteListEntry().req.headers };
    delete headers["host"];
    delete headers["content-length"];
    delete headers["accept-encoding"];
    return headers;
  }

  availableDates(): string[] {
    const dates: string[] = [];
    for (const entry of this.entries) {
      const path = entry.path || "";
      if (path.startsWith(CALENDAR_PREFIX) && entry.res) {
        const payload = decodeJsonBody(entry.res);
        const data = payload?.data;
        if (data?.list) {
          for (const item of data.list) {
            if (item.date) dates.push(String(item.date));
          }
        }
      }
      if (path.startsWith(SITE_LIST_PREFIX)) {
        const url = new URL(path, "http://x");
        const date = url.searchParams.get("date");
        if (date) dates.push(decodeURIComponent(date));
      }
    }
    const submitDate = String(this.submitBody().venues_date || "");
    if (submitDate) dates.push(submitDate);
    return unique(dates);
  }

  venueSnapshot(): VenueSnapshot {
    const body = this.submitBody();
    const courts = FIXED_COURTS;
    const submitDate = String(body.venues_date || "");
    const selectedTimes: TimeSlot[] = (body.venues_site_time || []).map((item: any) => ({
      start_time: String(item.start_time || ""),
      end_time: String(item.end_time || ""),
      start_timestamp: Number(item.start_timestamp || 0),
      end_timestamp: Number(item.end_timestamp || 0),
      price: String(item.price || "0"),
      times: String(item.times || "1"),
      source_date: String(item.source_date || submitDate),
    }));
    const selectedSiteId = body.venues_site_time?.[0]?.site_id ? Number(body.venues_site_time[0].site_id) : null;
    const selectedSiteName = body.venues_site_time?.[0]?.site_name ? String(body.venues_site_time[0].site_name) : "";
    const times: TimeSlot[] = FIXED_TIME_SLOTS.map((t) => ({ ...t, source_date: FIXED_SOURCE_DATE }));

    return {
      venues_id: String(body.venues_id || ""),
      date: submitDate,
      dates: this.availableDates(),
      courts,
      times,
      selected_site_id: selectedSiteId,
      selected_site_name: selectedSiteName,
      fixed_courts: selectedSiteId && selectedSiteName ? [{ site_id: selectedSiteId, site_name: selectedSiteName }] : [],
      selected_times: selectedTimes,
    };
  }

  buildSiteListRequest(params: Partial<Params>, date: string): RequestData {
    const entry = this.siteListEntry();
    const headers = { ...this.siteListHeaders(), ...(params.headers || {}) };
    const parsed = new URL(entry.url);
    parsed.searchParams.set("date", date);
    return { method: entry.req.method || "GET", url: parsed.toString(), headers, body: null };
  }

  buildSubmitRequest(params: Partial<Params> & { date?: string; court?: CourtOption; time_slots?: TimeSlot[] }): RequestData {
    const body = { ...this.submitBody() };
    const headers = { ...this.submitHeaders(), ...(params.headers || {}) };
    const venuesDate = params.date || body.venues_date;
    const court = params.court || firstCourt(params as any) || ({} as any);
    let timeSlots = (params.time_slots || []).slice(0, 2);
    if (!timeSlots.length) timeSlots = (body.venues_site_time || []).slice(0, 2);

    const siteId = court.site_id || (timeSlots[0] as any)?.site_id;
    const siteName = court.site_name || (timeSlots[0] as any)?.site_name;

    body.venues_date = venuesDate;
    body.venues_site_time = timeSlots.map((slot: any) => ({
      site_id: Number(slot.site_id || siteId),
      site_name: String(slot.site_name || siteName),
      start_time: String(slot.start_time),
      start_timestamp: shiftedTimestamp(slot, venuesDate, "start_timestamp"),
      end_timestamp: shiftedTimestamp(slot, venuesDate, "end_timestamp"),
      end_time: String(slot.end_time),
      times: String(slot.times || "1"),
      price: String(slot.price || "0"),
    }));

    return {
      method: this.submitEntry().req.method || "POST",
      url: this.submitEntry().url,
      headers,
      body,
    };
  }

  buildSubmitRequests(params: Partial<Params>): RequestData[] {
    const selections = (params as any).selections || [];
    if (selections.length) return this.buildSelectionRequests(params, selections);

    let courts = (params as any).courts || [];
    if (!courts.length && (params as any).court) courts = [(params as any).court];
    if (!courts.length) courts = [firstCourt(params as any)].filter(Boolean);
    let dates = params.dates || [];
    if (!dates.length && params.date) dates = [params.date];
    if (!dates.length) dates = [String(this.submitBody().venues_date || "")];

    const requests: RequestData[] = [];
    for (const date of dates) {
      for (const court of courts) {
        if (court) requests.push(this.buildSubmitRequest({ ...params, date, court } as any));
      }
    }
    return requests;
  }

  private buildSelectionRequests(params: Partial<Params>, selections: any[]): RequestData[] {
    let dates = params.dates || [];
    if (!dates.length && params.date) dates = [params.date];
    if (!dates.length) dates = [String(this.submitBody().venues_date || "")];

    const mode = (params as any).request_mode || "single";
    const groups = selectionGroups(selections, mode);
    const requests: RequestData[] = [];
    for (const date of dates) {
      for (const group of groups) {
        const court = group[0].court;
        const timeSlots = group.map((item: any) => item.time_slot);
        requests.push(this.buildSubmitRequest({ ...params, date, court, time_slots: timeSlots } as any));
      }
    }
    return requests;
  }
}

function decodeJsonBody(section: { base64?: string }): Record<string, any> {
  const encoded = section.base64;
  if (!encoded) return {};
  const decoded = atob(encoded);
  const bytes = new Uint8Array(decoded.length);
  for (let i = 0; i < decoded.length; i++) bytes[i] = decoded.charCodeAt(i);
  const text = new TextDecoder().decode(bytes);
  return JSON.parse(text);
}

function shiftedTimestamp(slot: any, targetDate: string, key: string): number {
  const original = Number(slot[key]);
  const sourceDate = String(slot.source_date || slot.date || "");
  if (!sourceDate) return original;
  try {
    const source = parseDate(sourceDate);
    const target = parseDate(targetDate);
    return original + Math.floor((target.getTime() - source.getTime()) / 1000);
  } catch {
    return original;
  }
}

function firstCourt(params: any): CourtOption | null {
  const courts = params.courts || [];
  if (courts.length) return courts[0];
  return params.court || null;
}

function selectionGroups(selections: any[], mode: string): any[][] {
  const normalized = selections
    .filter((item: any) => item.court && item.time_slot)
    .sort((a: any, b: any) => {
      const aKey = `${a.court.site_id}-${a.time_slot.start_timestamp}`;
      const bKey = `${b.court.site_id}-${b.time_slot.start_timestamp}`;
      return aKey.localeCompare(bKey);
    });

  if (mode !== "pair") return normalized.map((item: any) => [item]);

  const groups: any[][] = [];
  const used = new Set<number>();
  for (let i = 0; i < normalized.length; i++) {
    if (used.has(i)) continue;
    used.add(i);
    const pair = [normalized[i]];
    for (let j = i + 1; j < normalized.length; j++) {
      if (used.has(j)) continue;
      const sameCourt = String(normalized[i].court.site_id) === String(normalized[j].court.site_id);
      const adjacent = Number(normalized[i].time_slot.end_timestamp) === Number(normalized[j].time_slot.start_timestamp);
      if (sameCourt && adjacent) {
        pair.push(normalized[j]);
        used.add(j);
        break;
      }
    }
    groups.push(pair);
  }
  return groups;
}

function unique(values: string[]): string[] {
  const seen = new Set<string>();
  const result: string[] = [];
  for (const v of values) {
    if (v && !seen.has(v)) {
      seen.add(v);
      result.push(v);
    }
  }
  return result;
}

export function requestSummary(requestData: RequestData): Record<string, any> {
  const headers = { ...requestData.headers };
  if (headers["wx-token"]) {
    const v = headers["wx-token"];
    headers["wx-token"] = v.length <= 8 ? "*".repeat(v.length) : `${v.slice(0, 4)}...${v.slice(-4)}`;
  }
  return { method: requestData.method, url: requestData.url, headers, body: requestData.body };
}

export function siteListPayloadToDict(payload: any, date: string): { date: string; items: any[] } {
  if (!payload || typeof payload !== "object") return { date: decodeURIComponent(date || ""), items: [] };
  const data = payload.data || {};
  const items: any[] = [];
  for (const court of data.list || []) {
    const courtInfo = { site_id: court.site_id, site_name: court.site_name };
    for (const slot of court.site_data || []) {
      items.push({
        court: courtInfo,
        time_slot: {
          start_time: String(slot.start_time || ""),
          end_time: String(slot.end_time || ""),
          start_timestamp: Number(slot.start_timestamp || 0),
          end_timestamp: Number(slot.end_timestamp || 0),
          price: String(slot.price || "0"),
          times: String(slot.times || "1"),
          source_date: date,
        },
        status: slot.status,
        available: String(slot.status) === "2" && String(slot.times || "0") !== "0",
        disabled_desc: String(slot.disabled_desc || ""),
        disabled_reason: String(slot.disabled_reason || ""),
        member_name: String(slot.member_name || ""),
        mobile: String(slot.mobile || ""),
      });
    }
  }
  return { date: decodeURIComponent(date || ""), items };
}

File: badminton_booker/capture.py
from __future__ import annotations

import copy
import base64
import json
from dataclasses import dataclass
from datetime import datetime
from pathlib import Path
from urllib.parse import parse_qs, urlencode, unquote, urlparse, urlunparse

from .venue_defaults import FIXED_COURTS, FIXED_TIME_SLOTS, default_date


SUBMIT_PATH = "/v2/reserve/submit?"
SITE_LIST_PREFIX = "/v1/venues/venues_site_list"
CALENDAR_PREFIX = "/v1/venues/calendar"


@dataclass(frozen=True)
class CourtOption:
    site_id: int
    site_name: str


@dataclass(frozen=True)
class TimeOption:
    start_time: str
    end_time: str
    start_timestamp: int
    end_timestamp: int
    price: str
    times: str
    source_date: str = ""


@dataclass(frozen=True)
class VenueSnapshot:
    venues_id: str
    date: str
    dates: list[str]
    courts: list[CourtOption]
    times: list[TimeOption]
    selected_site_id: int | None
    selected_site_name: str
    fixed_courts: list[CourtOption]
    selected_times: list[TimeOption]


class CaptureStore:
    def __init__(self, path: str | Path = "request.txt", submit_path: str = SUBMIT_PATH) -> None:
        self.path = Path(path)
        self.submit_path = submit_path
        self.entries = json.loads(self.path.read_text(encoding="utf-8"))

    def submit_entry(self) -> dict:
        for entry in self.entries:
            if entry.get("path") == self.submit_path:
                return entry
        raise ValueError(f"No request found for path: {self.submit_path}")

    def submit_body(self) -> dict:
        return decode_json_body(self.submit_entry().get("req", {}))

    def submit_headers(self) -> dict[str, str]:
        headers = dict(self.submit_entry().get("req", {}).get("headers", {}))
        headers = sanitize_headers(headers)
        headers["content-type"] = "application/json"
        return headers

    def site_list_entry(self) -> dict:
        entry = self.latest_site_list_entry()
        if not entry:
            raise ValueError("No venues_site_list request found")
        return entry

    def site_list_headers(self) -> dict[str, str]:
        headers = dict(self.site_list_entry().get("req", {}).get("headers", {}))
        return sanitize_headers(headers)

    def build_site_list_request(self, params: dict, date: str) -> dict:
        entry = self.site_list_entry()
        headers = self.site_list_headers()
        headers.update(params.get("headers") or {})

        parsed = urlparse(str(entry.get("url") or ""))
        query = parse_qs(parsed.query)
        query["date"] = [date]
        url = urlunparse(parsed._replace(query=urlencode(query, doseq=True)))
        return {
            "method": entry.get("req", {}).get("method", "GET"),
            "url": url,
            "headers": headers,
            "body": None,
        }

    def latest_site_list_entry(self) -> dict | None:
        site_entries = [
            entry
            for entry in self.entries
            if str(entry.get("path", "")).startswith(SITE_LIST_PREFIX)
            and entry.get("hostname") == "stmember.styd.cn"
        ]
        if not site_entries:
            return None
        return max(site_entries, key=lambda entry: int(entry.get("order") or 0))

    def decode_site_list_response(self, entry: dict | None = None) -> dict:
        target = entry or self.site_list_entry()
        return decode_json_body(target.get("res", {}))

    def available_dates(self) -> list[str]:
        dates: list[str] = []
        for entry in self.entries:
            path = str(entry.get("path", ""))
            if path.startswith(CALENDAR_PREFIX):
                payload = decode_json_body(entry.get("res", {}))
                data = payload.get("data", {}) if isinstance(payload, dict) else {}
                for item in data.get("list", []):
                    date = item.get("date")
                    if date:
                        dates.append(str(date))
            if path.startswith(SITE_LIST_PREFIX):
                parsed = urlparse(path)
                date = parse_qs(parsed.query).get("date", [None])[0]
                if date:
                    dates.append(unquote(date))

        submit_date = str(self.submit_body().get("venues_date", ""))
        if submit_date:
            dates.append(submit_date)
        return _unique(dates)

    def venue_snapshot(self) -> VenueSnapshot:
        body = self.submit_body()
        courts = [
            CourtOption(site_id=int(item["site_id"]), site_name=str(item["site_name"])) for item in FIXED_COURTS
        ]

        submit_date = str(body.get("venues_date", ""))
        selected_times = [_time_from_submit_item(item, submit_date) for item in body.get("venues_site_time", [])]
        selected_site_id = selected_times and int(body["venues_site_time"][0].get("site_id")) or None
        selected_site_name = ""
        if body.get("venues_site_time"):
            selected_site_name = str(body["venues_site_time"][0].get("site_name", ""))

        times = [_time_from_fixed_item(item) for item in FIXED_TIME_SLOTS]

        return VenueSnapshot(
            venues_id=str(body.get("venues_id", "")),
            date=str(body.get("venues_date", "")),
            dates=self.available_dates(),
            courts=courts,
            times=times,
            selected_site_id=selected_site_id,
            selected_site_name=selected_site_name,
            fixed_courts=[
                CourtOption(site_id=selected_site_id, site_name=selected_site_name)
            ]
            if selected_site_id and selected_site_name
            else [],
            selected_times=selected_times,
        )

    def build_submit_request(self, params: dict) -> dict:
        body = copy.deepcopy(self.submit_body())
        headers = self.submit_headers()
        headers.update(params.get("headers") or {})

        venues_date = params.get("date") or body.get("venues_date")
        court = params.get("court") or _first_court(params) or {}
        time_slots = _limited_time_slots(params.get("time_slots") or [])
        if not time_slots:
            time_slots = _limited_time_slots(body.get("venues_site_time", []))

        site_id = court.get("site_id") or _first(time_slots, "site_id")
        site_name = court.get("site_name") or _first(time_slots, "site_name")

        body["venues_date"] = venues_date
        body["venues_site_time"] = [
            {
                "site_id": int(slot.get("site_id") or site_id),
                "site_name": str(slot.get("site_name") or site_name),
                "start_time": str(slot["start_time"]),
                "start_timestamp": shifted_timestamp(slot, venues_date, "start_timestamp"),
                "end_timestamp": shifted_timestamp(slot, venues_date, "end_timestamp"),
                "end_time": str(slot["end_time"]),
                "times": str(slot.get("times", "1")),
                "price": str(slot.get("price", "0")),
            }
            for slot in time_slots
        ]
        return {
            "method": self.submit_entry().get("req", {}).get("method", "POST"),
            "url": self.submit_entry().get("url"),
            "headers": headers,
            "body": body,
        }

    def build_submit_requests(self, params: dict) -> list[dict]:
        selections = params.get("selections") or []
        if selections:
            return self._build_selection_requests(params, selections)

        courts = params.get("courts") or []
        if not courts and params.get("court"):
            courts = [params["court"]]
        if not courts:
            courts = [_first_court(params)]
        dates = params.get("dates") or []
        if not dates and params.get("date"):
            dates = [params["date"]]
        if not dates:
            dates = [self.submit_body().get("venues_date", "") or default_date()]

        requests = []
        for date in dates:
            for court in courts:
                if court:
                    requests.append(self.build_submit_request({**params, "date": date, "court": court}))
        return requests

    def _build_selection_requests(self, params: dict, selections: list[dict]) -> list[dict]:
        dates = params.get("dates") or []
        if not dates and params.get("date"):
            dates = [params["date"]]
        if not dates:
            dates = [self.submit_body().get("venues_date", "") or default_date()]

        mode = params.get("request_mode", "single")
        groups = _selection_groups(selections, mode)
        requests = []
        for date in dates:
            for group in groups:
                court = group[0]["court"]
                time_slots = [item["time_slot"] for item in group]
                requests.append(self.build_submit_request({**params, "date": date, "court": court, "time_slots": time_slots}))
        return requests


def decode_json_body(section: dict) -> dict:
    encoded = section.get("base64")
    if not encoded:
        return {}
    decoded = base64.b64decode(encoded).decode("utf-8")
    return json.loads(decoded)


def sanitize_headers(headers: dict[str, str]) -> dict[str, str]:
    blocked = {"host", "content-length", "accept-encoding", "connection", "priority"}
    return {
        key: value
        for key, value in headers.items()
        if key
        and value
        and not key.startswith(":")
        and key.lower() not in blocked
    }


def snapshot_to_dict(snapshot: VenueSnapshot) -> dict:
    return {
        "venues_id": snapshot.venues_id,
        "date": snapshot.date,
        "dates": snapshot.dates,
        "courts": [court.__dict__ for court in snapshot.courts],
        "times": [time.__dict__ for time in snapshot.times],
        "selected_site_id": snapshot.selected_site_id,
        "selected_site_name": snapshot.selected_site_name,
        "fixed_courts": [court.__dict__ for court in snapshot.fixed_courts],
        "selected_times": [time.__dict__ for time in snapshot.selected_times],
    }


def site_list_payload_to_dict(payload: dict | None, date: str = "") -> dict:
    if not isinstance(payload, dict):
        return {"date": unquote(str(date or "")), "items": []}
    date = unquote(str(date or ""))
    data = payload.get("data", {}) if isinstance(payload, dict) else {}
    items = []
    for court in data.get("list", []):
        court_info = {"site_id": court.get("site_id"), "site_name": court.get("site_name")}
        for slot in court.get("site_data", []):
            items.append(
                {
                    "court": court_info,
                    "time_slot": {
                        "start_time": str(slot.get("start_time", "")),
                        "end_time": str(slot.get("end_time", "")),
                        "start_timestamp": int(slot.get("start_timestamp", 0)),
                        "end_timestamp": int(slot.get("end_timestamp", 0)),
                        "price": str(slot.get("price", "0")),
                        "times": str(slot.get("times", "1")),
                        "source_date": date,
                    },
                    "status": slot.get("status"),
                    "available": str(slot.get("status")) == "2" and str(slot.get("times", "0")) != "0",
                    "disabled_desc": str(slot.get("disabled_desc", "")),
                    "disabled_reason": str(slot.get("disabled_reason", "")),
                    "member_name": str(slot.get("member_name", "")),
                    "mobile": str(slot.get("mobile", "")),
                }
            )
    return {"date": date, "items": items}


def site_list_snapshot_to_dict(entry: dict | None) -> dict:
    if not entry:
        return {"date": "", "items": []}
    parsed = urlparse(str(entry.get("path", "")))
    date = parse_qs(parsed.query).get("date", [""])[0]
    payload = decode_json_body(entry.get("res", {}))
    return site_list_payload_to_dict(payload, date)


def request_summary(request_data: dict) -> dict:
    headers = dict(request_data.get("headers") or {})
    if "wx-token" in headers and headers["wx-token"]:
        headers["wx-token"] = mask_secret(headers["wx-token"])
    return {
        "method": request_data.get("method"),
        "url": request_data.get("url"),
        "headers": headers,
        "body": request_data.get("body"),
    }


def mask_secret(value: str) -> str:
    if len(value) <= 8:
        return "*" * len(value)
    return f"{value[:4]}...{value[-4:]}"


def _collect_times_for_site(site_data: dict, selected_site_id: int | None, source_date: str) -> list[TimeOption]:
    for court in site_data.get("list", []):
        if selected_site_id is not None and int(court.get("site_id", 0)) != selected_site_id:
            continue
        return [_time_from_site_item(item, source_date) for item in court.get("site_data", [])]
    return []


def _time_from_site_item(item: dict, source_date: str) -> TimeOption:
    return TimeOption(
        start_time=str(item.get("start_time", "")),
        end_time=str(item.get("end_time", "")),
        start_timestamp=int(item.get("start_timestamp", 0)),
        end_timestamp=int(item.get("end_timestamp", 0)),
        price=str(item.get("price", "0")),
        times=str(item.get("times", "1")),
        source_date=source_date,
    )


def _time_from_fixed_item(item: dict) -> TimeOption:
    return TimeOption(
        start_time=str(item.get("start_time", "")),
        end_time=str(item.get("end_time", "")),
        start_timestamp=int(item.get("start_timestamp", 0)),
        end_timestamp=int(item.get("end_timestamp", 0)),
        price=str(item.get("price", "0")),
        times=str(item.get("times", "1")),
        source_date=default_date(),
    )


def _time_from_submit_item(item: dict, source_date: str) -> TimeOption:
    return TimeOption(
        start_time=str(item.get("start_time", "")),
        end_time=str(item.get("end_time", "")),
        start_timestamp=int(item.get("start_timestamp", 0)),
        end_timestamp=int(item.get("end_timestamp", 0)),
        price=str(item.get("price", "0")),
        times=str(item.get("times", "1")),
        source_date=str(item.get("source_date", source_date)),
    )


def _first(items: list[dict], key: str) -> object:
    if not items:
        return None
    return items[0].get(key)


def _first_court(params: dict) -> dict | None:
    courts = params.get("courts") or []
    if courts:
        return courts[0]
    return params.get("court")


def _limited_time_slots(slots: list[dict]) -> list[dict]:
    return list(slots)[:2]


def _selection_groups(selections: list[dict], mode: str) -> list[list[dict]]:
    normalized = [item for item in selections if item.get("court") and item.get("time_slot")]
    normalized.sort(
        key=lambda item: (
            str(item["court"].get("site_id", "")),
            int(item["time_slot"].get("start_timestamp", 0)),
        )
    )
    if mode != "pair":
        return [[item] for item in normalized]

    groups: list[list[dict]] = []
    used: set[int] = set()
    for index, item in enumerate(normalized):
        if index in used:
            continue
        used.add(index)
        pair = [item]
        for other_index, other in enumerate(normalized):
            if other_index in used:
                continue
            same_court = str(item["court"].get("site_id")) == str(other["court"].get("site_id"))
            adjacent = int(item["time_slot"].get("end_timestamp", 0)) == int(other["time_slot"].get("start_timestamp", -1))
            if same_court and adjacent:
                pair.append(other)
                used.add(other_index)
                break
        groups.append(pair)
    return groups


def shifted_timestamp(slot: dict, target_date: str, key: str) -> int:
    original = int(slot[key])
    source_date = str(slot.get("source_date") or slot.get("date") or "")
    if not source_date:
        return original
    try:
        source = _parse_date(source_date)
        target = _parse_date(target_date)
    except ValueError:
        return original
    return original + int((target - source).total_seconds())


def _parse_date(value: str) -> datetime:
    normalized = value.replace("-", "/")
    return datetime.strptime(normalized, "%Y/%m/%d")


def _unique(values: list[str]) -> list[str]:
    seen = set()
    result = []
    for value in values:
        if value and value not in seen:
            seen.add(value)
            result.append(value)
    return result


def _site_entry_date(entry: dict | None) -> str:
    if not entry:
        return ""
    parsed = urlparse(str(entry.get("path", "")))
    return parse_qs(parsed.query).get("date", [""])[0]
